src/main/java/com/google/devtools/build/lib/concurrent/MultiExecutorQueueVisitor.java

```diff
@@ -0,0 +1,125 @@
+      ExecutorService cpuHeavyPoolExecutorService,
+      ExecutorService executionPhaseExecutorService,
+      boolean failFastOnException,
+      ErrorClassifier errorClassifier) {
+    super(
+        regularPoolExecutorService,
+        /*shutdownOnCompletion=*/ true,
+        failFastOnException,
+        errorClassifier);
+    this.regularPoolExecutorService = super.getExecutorService();
+    this.cpuHeavyPoolExecutorService = Preconditions.checkNotNull(cpuHeavyPoolExecutorService);
+    this.executionPhaseExecutorService = executionPhaseExecutorService;
+  }
+
+  public static MultiExecutorQueueVisitor createWithExecutorServices(
+      ExecutorService regularPoolExecutorService,
+      ExecutorService cpuHeavyPoolExecutorService,
+      boolean failFastOnException,
+      ErrorClassifier errorClassifier) {
+    return createWithExecutorServices(
+        regularPoolExecutorService,
+        cpuHeavyPoolExecutorService,
+        /*executionPhaseExecutorService=*/ null,
+        failFastOnException,
+        errorClassifier);
+  }
+
+  public static MultiExecutorQueueVisitor createWithExecutorServices(
+      ExecutorService regularPoolExecutorService,
+      ExecutorService cpuHeavyPoolExecutorService,
+      ExecutorService executionPhaseExecutorService,
+      boolean failFastOnException,
+      ErrorClassifier errorClassifier) {
+    return new MultiExecutorQueueVisitor(
+        regularPoolExecutorService,
+        cpuHeavyPoolExecutorService,
+        executionPhaseExecutorService,
+        failFastOnException,
+        errorClassifier);
+  }
+
+  @Override
+  public final void execute(Runnable runnable, ThreadPoolType threadPoolType) {
+    super.executeWithExecutorService(runnable, getExecutorServiceByThreadPoolType(threadPoolType));
+  }
+
+  @VisibleForTesting
+  ExecutorService getExecutorServiceByThreadPoolType(ThreadPoolType threadPoolType) {
+    switch (threadPoolType) {
+      case REGULAR:
+        return regularPoolExecutorService;
+      case CPU_HEAVY:
+        return cpuHeavyPoolExecutorService;
+      case EXECUTION_PHASE:
+        Preconditions.checkNotNull(executionPhaseExecutorService);
+        return executionPhaseExecutorService;
+    }
+    throw new IllegalStateException("Invalid ThreadPoolType: " + threadPoolType);
+  }
+
+  @Override
+  protected void shutdownExecutorService(Throwable catastrophe) {
+    if (catastrophe != null) {
+      Throwables.throwIfUnchecked(catastrophe);
+    }
+    internalShutdownExecutorService(regularPoolExecutorService);
+    internalShutdownExecutorService(cpuHeavyPoolExecutorService);
+    if (executionPhaseExecutorService != null) {
+      internalShutdownExecutorService(executionPhaseExecutorService);
+    }
+  }
+
+  private void internalShutdownExecutorService(ExecutorService executorService) {
+    executorService.shutdown();
+    while (true) {
+      try {
+        executorService.awaitTermination(Integer.MAX_VALUE, SECONDS);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:** 
The loop containing this operation may iterate indefinitely if the operation throws an exception on multiple instances of its evaluation. As the operation throws an exception, it bypasses exit points in the loop and continues to the next iteration through a catch clause.

**Fix:** 
Add a retry counter to ensure that the loop terminates after a predefined number of iterations.